### PR TITLE
SNOW-704094 Flatten select after filter

### PR DESCRIPTION
**Pull request by sfc-gh-stan**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-704094

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   The code change checks if the subquery's where clause depends on a column that is **NEW**  in the current query, if it does not, flatten the query. Note that, `WHERE` is evaluated on the subquery first, so it is safe to flatten when there are columns that are **CHANGED** between the current query and the subquery.

### CODE REVIEW

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py [hunk 1]
```diff
@@ -47,6 +49,18 @@
 SET_EXCEPT = analyzer_utils.EXCEPT
 
 
+@dataclass
+class WhereClause:
+    expression: Expression
+    dependent_columns: Set[str]
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/_internal/analyzer/select_statement.py:55]: Should be `Optional[Set[str]]` since depdent_columns can be `None` when the Expression is not parsed (e.g. created by `sql_expr`)

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py [hunk 2]
```diff
@@ -47,6 +49,18 @@
...
+
+
+@dataclass
+class OrderByClause:
+    expressions: List[Expression]
+    dependent_columns: Set[str]
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/_internal/analyzer/select_statement.py:61]: Ditto.

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py [hunk 3]
```diff
@@ -582,11 +595,13 @@ def sort(self, cols: List[Expression]) -> "SelectStatement":
             new.from_ = self.from_.to_subqueryable()
             new.pre_actions = new.from_.pre_actions
             new.post_actions = new.from_.post_actions
-            new.order_by = cols
+            new.order_by = OrderByClause(cols, dependent_columns)
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/_internal/analyzer/select_statement.py:598]: This a is a bug that will be addressed by [SNOW-756443](https://snowflakecomputing.atlassian.net/browse/SNOW-756443), where chained sorts/orderby should preserve all the dependent columns in order.

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py (hunk 1 above)
Comment (sfc-gh-sfan) [reply src/snowflake/snowpark/_internal/analyzer/select_statement.py:55]: Should we make the change then?

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py [hunk 4]
```diff
@@ -488,49 +501,41 @@ def select(self, cols: List[Expression]) -> "SelectStatement":
...
             can_be_flattened = False
             disable_next_level_flatten = True
-        elif self.flatten_disabled or self.has_clause_using_columns:
+        elif self.flatten_disabled:
+            can_be_flattened = False
+        elif (
+            self.where
+            and self.where.dependent_columns
+            and any(
+                new_column_states[_col].change_state in (ColumnChangeState.NEW,)
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/_internal/analyzer/select_statement.py:510]: nit: use `==` instead?

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py [hunk 5]
```diff
@@ -488,49 +501,41 @@ def select(self, cols: List[Expression]) -> "SelectStatement":
...
+                for _col in self.where.dependent_columns.intersection(
+                    new_column_states.active_columns
+                )
+            )
+        ):
+            can_be_flattened = False
+        elif self.order_by:  # TODO: implement orderby
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/_internal/analyzer/select_statement.py:517]: Not sure about the policy in open source repo but might be good to have a JIRA?

**Review by sfc-gh-sfan — COMMENTED** [review]
There are some neat refactors though it also makes the change hard to parse. Could you help highlight the core change?

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py (hunk 5 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/_internal/analyzer/select_statement.py:517]: SNOW-704094 covers both filter and orderby, I just want to separate the PR's for easier review.

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py [hunk 6]
```diff
@@ -549,31 +554,39 @@ def select(self, cols: List[Expression]) -> "SelectStatement":
...
+            dependent_columns, self.column_states
+        )
         if can_be_flattened:
             new = copy(self)
             new.from_ = self.from_.to_subqueryable()
             new.pre_actions = new.from_.pre_actions
             new.post_actions = new.from_.post_actions
-            new.where = And(self.where, col) if self.where is not None else col
             new._column_states = self._column_states
+            if self.where:
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/_internal/analyzer/select_statement.py:569]: Core change #1: merge where clauses between parent query and child query by AND the condition expressions, track the dependent columns of a WHERE clause, merge two dependent columns sets using UNION.
@sfc-gh-sfan @sfc-gh-yixie

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py [hunk 7]
```diff
@@ -488,49 +501,41 @@ def select(self, cols: List[Expression]) -> "SelectStatement":
             # We don't flatten when there are duplicate columns.
             can_be_flattened = False
             disable_next_level_flatten = True
-        elif self.flatten_disabled or self.has_clause_using_columns:
+        elif self.flatten_disabled:
+            can_be_flattened = False
+        elif (
+            self.where
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/_internal/analyzer/select_statement.py:507]: Core change #2: Previously, as long as subquery has a WHERE clause, `can_be_flattened` evaluates to `False`. This change deems that the only case not to be flattened is when any dependent column from the WHERE clause is newly added in the current query, i.e. `df.filter(col("C") > 10).select("*", col(Literal(12)).alias("C")` should not be flattened to `SELECT *, 12 as C FROM <subquery> WHERE C > 10`.

Comment (sfc-gh-sfan) [reply src/snowflake/snowpark/_internal/analyzer/select_statement.py:509]: This is due to my lack of knowledge of the simplifier: Do we handle the case where select added alias to existing column? E.g. `df.filter(col("C") > 10).select(col("A").alias("C"))`? Is this alias also considered a new column? If so then I think it should be okay.

Comment (sfc-gh-stan) [reply src/snowflake/snowpark/_internal/analyzer/select_statement.py:509]: We don't flatten `df.filter(col("C") > 10).select(col("A").alias("C"))`. `C` is considered a new column in this case.

**Review by sfc-gh-sfan — APPROVED** [review]
LGTM. Let's also have @sfc-gh-yixie take a look

File: tests/integ/test_simplifier_suite.py [hunk 8]
```diff
@@ -995,6 +996,46 @@ def test_rename_to_existing_column_column(session):
     Utils.check_answer(df4, [Row(3, 3, 1)])
 
 
+@pytest.mark.parametrize(
+    "operation,simplified_query",
```
Comment (sfc-gh-yixie) [inline tests/integ/test_simplifier_suite.py:1000]: If I use sql expression, does it flatten, for instance `df.filter(sql_expr('a > 1')).select(...)`?
It shouldn't because we can't detect the dependent columns of a sql text.

Comment (sfc-gh-stan) [reply tests/integ/test_simplifier_suite.py:1000]: Good catch! I fixed and added a test case.
